File: ports/go/src/core/kernel/bcseventloop.go
package kernel

import (
	"sync"
)

// BcsEventLoop models QEventLoop
type BcsEventLoop struct {
	*BcsObject
	dispatcher *BcsEventDispatcher
	mu         sync.Mutex
	isRunning  bool
	quitCode   int
}

func NewBcsEventLoop(parent *BcsObject) *BcsEventLoop {
	return &BcsEventLoop{
		BcsObject:  NewBcsObject(parent),
		dispatcher: NewBcsEventDispatcher(),
		isRunning:  false,
		quitCode:   0,
	}
}

// Exec enters the main event loop and waits until exit() is called
func (l *BcsEventLoop) Exec() int {
	l.mu.Lock()
	if l.isRunning {
		l.mu.Unlock()
		return -1 // Already running
	}
	l.isRunning = true
	l.mu.Unlock()

	// Blocks until Stop is called
	l.dispatcher.Exec()

	l.mu.Lock()
	defer l.mu.Unlock()
	l.isRunning = false
	return l.quitCode
}

// Exit tells the event loop to exit with a return code
func (l *BcsEventLoop) Exit(returnCode int) {
	l.mu.Lock()
	defer l.mu.Unlock()

	l.quitCode = returnCode
	l.dispatcher.Stop()
}

// Quit is a convenience for exit(0)
func (l *BcsEventLoop) Quit() {
	l.Exit(0)
}

func (l *BcsEventLoop) IsRunning() bool {
	l.mu.Lock()
	defer l.mu.Unlock()
	return l.isRunning
}

File: ports/java/src/core/kernel/BcsEventLoop.java
package core.kernel;

import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicInteger;

public class BcsEventLoop extends BcsObject {
    private final BcsEventDispatcher dispatcher;

    public BcsEventDispatcher getDispatcher() {
        return dispatcher;
    }

    private final AtomicBoolean isRunning = new AtomicBoolean(false);
    private final AtomicInteger quitCode = new AtomicInteger(0);

    public BcsEventLoop() {
        this(new BcsEventDispatcher(), null);
    }

    public BcsEventLoop(BcsEventDispatcher dispatcher, BcsObject parent) {
        super(parent);
        this.dispatcher = dispatcher != null ? dispatcher : new BcsEventDispatcher();
    }

    public int exec() {
        if (!isRunning.compareAndSet(false, true)) {
            return -1;
        }

        dispatcher.exec();

        isRunning.set(false);
        return quitCode.get();
    }

    public void exit(int returnCode) {
        quitCode.set(returnCode);
        dispatcher.stop();
    }

    public void quit() {
        exit(0);
    }

    public boolean isRunning() {
        return isRunning.get();
    }
}
